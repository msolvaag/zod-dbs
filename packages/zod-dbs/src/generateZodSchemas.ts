import { logDebug } from 'zod-dbs-core';

import type {
  ZodDbsConfig,
  ZodDbsDbConnector,
  ZodDbsSchemaInfo,
} from 'zod-dbs-core';

import { DEFAULT_CONFIGURATION } from './constants.js';
import { generateConstantsFile } from './generate/generateConstantsFile.js';
import { generateIndexFiles } from './generate/generateIndexFiles.js';
import { generateSchemaFiles } from './generate/generateSchemaFile.js';
import { generateTypesFile } from './generate/generateTypesFile.js';
import { clearTablesDirectory } from './utils/index.js';

/**
 * Generates Zod schemas for all tables in the specified database schema.
 */
export const generateZodSchemas = async (
  dbConnector: ZodDbsDbConnector,
  config: ZodDbsConfig
): Promise<ZodDbsSchemaInfo> => {
  const generateConfig = {
    ...DEFAULT_CONFIGURATION,
    ...config,
  };

  const { connectionString, outputDir, schemaName, cleanOutput, onProgress } =
    generateConfig;

  if (cleanOutput) {
    clearTablesDirectory(outputDir);
  }

  logDebug(`Connecting to Postgres database at ${connectionString}`);

  const schema = await dbConnector.getSchemaInformation(config);

  onProgress?.('generating', { total: schema.tables.length });

  logDebug(
    `Generating zod schemas for ${schema.tables.length} tables in db schema '${schemaName}'`
  );

  for (const table of schema.tables) {
    await generateSchemaFiles(table, generateConfig);
  }

  await generateIndexFiles(schema, generateConfig);
  await generateConstantsFile(schema, generateConfig);
  await generateTypesFile(schema, generateConfig);

  onProgress?.('done');

  return schema;
};
