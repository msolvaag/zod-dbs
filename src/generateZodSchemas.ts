import { Client } from 'pg';

import type { ZodPgConfig, ZodPgProgress, ZodPgSchemaInfo } from './types.js';

import { createClient } from './database/client.js';
import { getSchemaInformation } from './database/schema.js';
import { generateConstantsFile } from './generate/generateConstantsFile.js';
import { generateIndexFiles } from './generate/generateIndexFiles.js';
import { generateSchemaFiles } from './generate/generateSchemaFile.js';
import { generateTypesFile } from './generate/generateTypesFile.js';
import { createTableModel } from './generate/models.js';
import { clearTablesDirectory, logDebug } from './utils/index.js';

export interface ZodPgGenerateConfig extends ZodPgConfig {
  outputDir: string;
  onProgress?: (status: ZodPgProgress, args?: unknown) => void;
}

const defaultConfig = {
  defaultEmptyArray: false,
  stringifyJson: true,
  zodVersion: 3,
  fieldNameCasing: 'camelCase',
  objectNameCasing: 'PascalCase',
  moduleResolution: 'commonjs',
} satisfies Partial<ZodPgGenerateConfig>;

/**
 * Generates Zod schemas for all tables in the specified Postgres database schema.
 */
export const generateZodSchemas = async ({
  onProgress,
  ...config
}: ZodPgGenerateConfig): Promise<ZodPgSchemaInfo> => {
  const generateConfig = {
    ...defaultConfig,
    ...config,
  };

  const { connection, outputDir, schemaName, include, exclude, cleanOutput } =
    generateConfig;

  if (cleanOutput) {
    clearTablesDirectory(outputDir);
  }

  let client: Client | undefined = undefined;

  try {
    logDebug(
      `Connecting to Postgres database at ${connection.connectionString}`
    );

    client = createClient(connection);

    onProgress?.('connecting');

    await client.connect();
    logDebug(`Connected to Postgres database`);

    onProgress?.('fetchingSchema');
    const schema = await getSchemaInformation(client, {
      schemaName,
      include,
      exclude,
    });

    onProgress?.('generating', { total: schema.tables.length });

    logDebug(
      `Generating zod schemas for ${schema.tables.length} tables in db schema '${schemaName}'`
    );

    const models = [];

    for (const tableInfo of schema.tables) {
      logDebug(`Generating model for: ${tableInfo.type} ${tableInfo.name}`);
      const model = await createTableModel(tableInfo, generateConfig);
      models.push(model);
    }

    for (const tableModel of models) {
      await generateSchemaFiles(tableModel, generateConfig);
    }

    await generateIndexFiles(models, generateConfig);
    await generateConstantsFile(schema, generateConfig);
    await generateTypesFile(schema, generateConfig);

    onProgress?.('done');

    return schema;
  } finally {
    await client?.end();
  }
};
