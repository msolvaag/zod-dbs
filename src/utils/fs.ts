import { existsSync, readdirSync, unlinkSync } from 'fs';
import { mkdir } from 'fs/promises';
import path from 'path';

import { ZodPgTableType } from '../types.js';
import { logDebug } from './debug.js';

/**
 * Deletes all .ts files in the output schemas folder.
 */
export function clearTablesDirectory(outputPath: string) {
  // Check if the directory exists first
  if (!existsSync(outputPath)) {
    logDebug(`Directory ${outputPath} does not exist, nothing to clear`);
    return;
  }

  const files = readdirSync(outputPath, {
    recursive: true,
    withFileTypes: true,
  });

  for (const file of files) {
    if (file.isDirectory()) continue;
    if (!file.name.endsWith('.ts')) continue;

    const filePath = path.join(file.path, file.name);
    unlinkSync(filePath);
  }

  if (files.length > 0) {
    logDebug(`Deleted all .ts files in ${outputPath}`);
  }
}

export async function ensureFolder(folderPath: string) {
  const exists = existsSync(folderPath);

  if (!exists) {
    await mkdir(folderPath, { recursive: true });
  }
}

export const getOutputFolder = (type: ZodPgTableType): string => {
  switch (type) {
    case 'table':
      return 'tables';
    case 'materialized_view':
      return 'materialized_views';
    case 'view':
      return 'views';
    case 'foreign_table':
      return 'foreign_tables';
    default:
      return 'others';
  }
};
