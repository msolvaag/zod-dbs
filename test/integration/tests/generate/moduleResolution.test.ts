import fs from 'fs';
import path from 'path';

import { generateZodSchemas } from '../../../../src/generateZodSchemas.js';
import {
  deleteOutputFiles,
  getClientConnectionString,
  getOutputFiles,
  setupTestDb,
  teardownTestDb,
  TestDbContext,
} from '../../testDbUtils.js';

let ctx: TestDbContext;

const outputDir = `${import.meta.dirname}/test-output/module-resolution`;
let connectionString: string;

beforeAll(async () => {
  ctx = await setupTestDb();
  connectionString = getClientConnectionString();
});

afterAll(async () => {
  await teardownTestDb(ctx);
  await deleteOutputFiles(outputDir);
});

describe('module resolution formats', () => {
  it('generates modules without file extensions (commonjs module resolution)', async () => {
    await generateZodSchemas({
      connection: {
        connectionString,
        ssl: false,
      },
      outputDir: `${outputDir}/commonjs`,
      moduleResolution: 'commonjs',
      include: ['users'],
    });

    const outputFiles = await getOutputFiles(`${outputDir}/commonjs`);

    for (const file of outputFiles) {
      const content = fs.readFileSync(file, 'utf8');

      // Check that imports/exports don't include file extensions
      const importExportLines = content
        .split('\n')
        .filter((line) => line.includes('import') || line.includes('export'));

      for (const line of importExportLines) {
        if (line.includes("from './") || line.includes("from '../")) {
          expect(line).not.toMatch(/\.js['"]|\.ts['"]/);
        }
      }

      expect(content).toMatchSnapshot(
        `commonjs/${path.relative(`${outputDir}/commonjs`, file)}`
      );
    }
  });

  it('generates modules with file extensions (esm module resolution)', async () => {
    await generateZodSchemas({
      connection: {
        connectionString,
        ssl: false,
      },
      outputDir: `${outputDir}/esm`,
      moduleResolution: 'esm',
      include: ['users'],
    });

    const outputFiles = await getOutputFiles(`${outputDir}/esm`);

    for (const file of outputFiles) {
      const content = fs.readFileSync(file, 'utf8');

      // Check that relative imports include file extensions
      const importExportLines = content
        .split('\n')
        .filter((line) => line.includes('import') || line.includes('export'));

      for (const line of importExportLines) {
        if (line.includes("from './") || line.includes("from '../")) {
          expect(line).toMatch(/\.js['"]/);
        }
      }

      expect(content).toMatchSnapshot(
        `esm/${path.relative(`${outputDir}/esm`, file)}`
      );
    }
  });
});
