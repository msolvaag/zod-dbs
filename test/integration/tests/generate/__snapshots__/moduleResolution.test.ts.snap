// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`module resolution formats > generates modules with file extensions (esm module resolution) > esm/constants.ts 1`] = `
"// This file is auto-generated by zod-pg. Do not edit this file directly.

export const TABLE_USERS = 'users';
"
`;

exports[`module resolution formats > generates modules with file extensions (esm module resolution) > esm/tables/index.ts 1`] = `
"// This file is auto-generated by zod-pg. Do not edit this file directly.

export * from './users/index.js';
"
`;

exports[`module resolution formats > generates modules with file extensions (esm module resolution) > esm/tables/users/index.ts 1`] = `
"// This file is auto-generated by zod-pg. Do not edit this file directly.

export * from './schema.js';
"
`;

exports[`module resolution formats > generates modules with file extensions (esm module resolution) > esm/tables/users/schema.ts 1`] = `
"// This file is auto-generated by zod-pg. Do not edit this file directly.

import { z } from 'zod';

export const USER_ROLES = ['admin', 'editor', 'viewer'] as const;

/**
 * The base read schema for the "public.users" table.
 * This schema is used to validate the data read from the database without any transformations.
 */
export const UsersTableReadSchema = z.object({
    /**
    * dataType: int4
    * defaultValue: nextval('users_id_seq'::regclass)
    */
    id: z.number().int(),
    /**
    * dataType: varchar
    * defaultValue: 
    */
    name: z.string(),
    /**
    * dataType: varchar
    * defaultValue: 
    */
    email: z.string().nullish().transform((value) => value ?? undefined).optional(),
    /**
    * dataType: varchar
    * defaultValue: 'active'::character varying
    */
    status: z.string().nullish().transform((value) => value ?? undefined).optional(),
    /**
    * dataType: jsonb
    * defaultValue: 
    */
    profile: z.any().nullish().transform((value) => value ?? undefined).optional(),
    /**
    * dataType: timestamptz
    * defaultValue: now()
    */
    created_at: z.date().nullish().transform((value) => value ?? undefined).optional(),
    /**
    * dataType: _text
    * defaultValue: 
    */
    roles: z.array(z.enum(USER_ROLES)).nullish().transform((value) => value ?? undefined).optional(),
    /**
    * dataType: _timestamp
    * defaultValue: 
    */
    dates: z.array(z.date()).nullish().transform((value) => value ?? undefined).optional(),
});

/**
 * The base record type for the "public.users" table.
 * This type represents the raw database record without any transformations.
 */
export type UserReadBaseRecord = z.output<typeof UsersTableReadSchema>;

/**
* The read transform function for the "public.users" table.
* Maps the raw database fields to expected property names. e.g snake_case to camelCase.
*/
export const transformUserReadRecord = (data: UserReadBaseRecord): {
    id: UserReadBaseRecord['id'],
    name: UserReadBaseRecord['name'],
    email?: UserReadBaseRecord['email'],
    status?: UserReadBaseRecord['status'],
    profile?: UserReadBaseRecord['profile'],
    createdAt?: UserReadBaseRecord['created_at'],
    roles?: UserReadBaseRecord['roles'],
    dates?: UserReadBaseRecord['dates'],
} => ({
    id: data.id,
    name: data.name,
    email: data.email,
    status: data.status,
    profile: data.profile,
    createdAt: data.created_at,
    roles: data.roles,
    dates: data.dates,
});

/**
 * The read schema for the "public.users" table.
 * This schema is used to validate the data read from the database with transformations.
 */
export const UsersTableSchema = UsersTableReadSchema.transform(transformUserReadRecord);

/**
 * The base write schema for the "public.users" table.
 * This schema is used to validate the data before writing to the database without any transformations.
 */
export const UsersTableWriteSchema = z.object({
    /**
    * dataType: varchar
    * defaultValue: 
    */
    name: z.string().max(100),
    /**
    * dataType: varchar
    * defaultValue: 
    */
    email: z.string().max(255).nullish().optional(),
    /**
    * dataType: varchar
    * defaultValue: 'active'::character varying
    */
    status: z.string().max(20).nullish().optional(),
    /**
    * dataType: jsonb
    * defaultValue: 
    */
    profile: z.any().nullish().transform((value) => value ? JSON.stringify(value) : value).optional(),
    /**
    * dataType: timestamptz
    * defaultValue: now()
    */
    createdAt: z.date().nullish().optional(),
    /**
    * dataType: _text
    * defaultValue: 
    */
    roles: z.array(z.enum(USER_ROLES)).nullish().optional(),
    /**
    * dataType: _timestamp
    * defaultValue: 
    */
    dates: z.array(z.date()).nullish().optional(),
});

/**
 * The base record type for the "public.users" table.
 * This type represents an insertable database record before casing transformations are applied.
 */
export type UserInsertBaseRecord = z.output<typeof UsersTableWriteSchema>;

/**
 * The base record type for the "public.users" table.
 * This type represents an updatable database record before casing transformations are applied.
 */
export type UserUpdateBaseRecord = Partial<UserInsertBaseRecord>;

/**
 * The insert transform function for the "public.users" table.
 * Maps the expected property names to raw database fields. e.g camelCase to snake_case.
 */
export const transformUserInsertRecord = (data: UserInsertBaseRecord): {
    name: UserInsertBaseRecord['name'],
    email?: UserInsertBaseRecord['email'],
    status?: UserInsertBaseRecord['status'],
    profile?: UserInsertBaseRecord['profile'],
    created_at?: UserInsertBaseRecord['createdAt'],
    roles?: UserInsertBaseRecord['roles'],
    dates?: UserInsertBaseRecord['dates'],
} => ({
    name: data.name,
    email: data.email,
    status: data.status,
    profile: data.profile,
    created_at: data.createdAt,
    roles: data.roles,
    dates: data.dates,
});

/**
 * The update transform function for the "public.users" table.
 * Maps the expected property names to raw database fields. e.g camelCase to snake_case.
 */
export const transformUserUpdateRecord = (data: UserUpdateBaseRecord): {
    name: UserUpdateBaseRecord['name'],
    email?: UserUpdateBaseRecord['email'],
    status?: UserUpdateBaseRecord['status'],
    profile?: UserUpdateBaseRecord['profile'],
    created_at?: UserUpdateBaseRecord['createdAt'],
    roles?: UserUpdateBaseRecord['roles'],
    dates?: UserUpdateBaseRecord['dates'],
} => ({
    name: data.name,
    email: data.email,
    status: data.status,
    profile: data.profile,
    created_at: data.createdAt,
    roles: data.roles,
    dates: data.dates,
});

/**
 * The insert schema for the "public.users" table.
 * This schema is used to validate and transform a record before inserting into the database.
 */
export const UsersTableInsertSchema = UsersTableWriteSchema.transform(transformUserInsertRecord);

/**
 * The update schema for the "public.users" table.
 * This schema is used to validate and transform a record before updating the database.
 */
export const UsersTableUpdateSchema = UsersTableWriteSchema.partial().transform(transformUserUpdateRecord);

type TableInsertRecord = z.input<typeof UsersTableInsertSchema>;
type TableReadRecord = z.output<typeof UsersTableSchema>;
export type UserRole = (typeof USER_ROLES)[number];

/**
* Represents a database record from the "public.users" table.
*/
export interface UserRecord {
    /**
    */
    id: TableReadRecord['id'];
    /**
    */
    name: TableReadRecord['name'];
    /**
    */
    email?: TableReadRecord['email'];
    /**
    */
    status?: TableReadRecord['status'];
    /**
    */
    profile?: TableReadRecord['profile'];
    /**
    */
    createdAt?: TableReadRecord['createdAt'];
    /**
    */
    roles?: TableReadRecord['roles'];
    /**
    */
    dates?: TableReadRecord['dates'];
}

/**
* Represents an insertable database record from the "public.users" table.
*/
export interface UserInsertRecord {
    /**
    * @maxLen: 100
    */
    name: TableInsertRecord['name'];
    /**
    * @maxLen: 255
    */
    email?: TableInsertRecord['email'];
    /**
    * @maxLen: 20
    * @default: 'active'::character varying
    */
    status?: TableInsertRecord['status'];
    /**
    */
    profile?: TableInsertRecord['profile'];
    /**
    * @default: now()
    */
    createdAt?: TableInsertRecord['createdAt'];
    /**
    */
    roles?: TableInsertRecord['roles'];
    /**
    */
    dates?: TableInsertRecord['dates'];
}

/**
* Represents an updateable database record from the "public.users" table.
*/
export type UserUpdateRecord = Partial<UserInsertRecord>;
"
`;

exports[`module resolution formats > generates modules with file extensions (esm module resolution) > esm/types.ts 1`] = `
"// This file is auto-generated by zod-pg. Do not edit this file directly.

export type Table = 
    | 'users'
"
`;

exports[`module resolution formats > generates modules without file extensions (commonjs module resolution) > commonjs/constants.ts 1`] = `
"// This file is auto-generated by zod-pg. Do not edit this file directly.

export const TABLE_USERS = 'users';
"
`;

exports[`module resolution formats > generates modules without file extensions (commonjs module resolution) > commonjs/tables/index.ts 1`] = `
"// This file is auto-generated by zod-pg. Do not edit this file directly.

export * from './users';
"
`;

exports[`module resolution formats > generates modules without file extensions (commonjs module resolution) > commonjs/tables/users/index.ts 1`] = `
"// This file is auto-generated by zod-pg. Do not edit this file directly.

export * from './schema';
"
`;

exports[`module resolution formats > generates modules without file extensions (commonjs module resolution) > commonjs/tables/users/schema.ts 1`] = `
"// This file is auto-generated by zod-pg. Do not edit this file directly.

import { z } from 'zod';

export const USER_ROLES = ['admin', 'editor', 'viewer'] as const;

/**
 * The base read schema for the "public.users" table.
 * This schema is used to validate the data read from the database without any transformations.
 */
export const UsersTableReadSchema = z.object({
    /**
    * dataType: int4
    * defaultValue: nextval('users_id_seq'::regclass)
    */
    id: z.number().int(),
    /**
    * dataType: varchar
    * defaultValue: 
    */
    name: z.string(),
    /**
    * dataType: varchar
    * defaultValue: 
    */
    email: z.string().nullish().transform((value) => value ?? undefined).optional(),
    /**
    * dataType: varchar
    * defaultValue: 'active'::character varying
    */
    status: z.string().nullish().transform((value) => value ?? undefined).optional(),
    /**
    * dataType: jsonb
    * defaultValue: 
    */
    profile: z.any().nullish().transform((value) => value ?? undefined).optional(),
    /**
    * dataType: timestamptz
    * defaultValue: now()
    */
    created_at: z.date().nullish().transform((value) => value ?? undefined).optional(),
    /**
    * dataType: _text
    * defaultValue: 
    */
    roles: z.array(z.enum(USER_ROLES)).nullish().transform((value) => value ?? undefined).optional(),
    /**
    * dataType: _timestamp
    * defaultValue: 
    */
    dates: z.array(z.date()).nullish().transform((value) => value ?? undefined).optional(),
});

/**
 * The base record type for the "public.users" table.
 * This type represents the raw database record without any transformations.
 */
export type UserReadBaseRecord = z.output<typeof UsersTableReadSchema>;

/**
* The read transform function for the "public.users" table.
* Maps the raw database fields to expected property names. e.g snake_case to camelCase.
*/
export const transformUserReadRecord = (data: UserReadBaseRecord): {
    id: UserReadBaseRecord['id'],
    name: UserReadBaseRecord['name'],
    email?: UserReadBaseRecord['email'],
    status?: UserReadBaseRecord['status'],
    profile?: UserReadBaseRecord['profile'],
    createdAt?: UserReadBaseRecord['created_at'],
    roles?: UserReadBaseRecord['roles'],
    dates?: UserReadBaseRecord['dates'],
} => ({
    id: data.id,
    name: data.name,
    email: data.email,
    status: data.status,
    profile: data.profile,
    createdAt: data.created_at,
    roles: data.roles,
    dates: data.dates,
});

/**
 * The read schema for the "public.users" table.
 * This schema is used to validate the data read from the database with transformations.
 */
export const UsersTableSchema = UsersTableReadSchema.transform(transformUserReadRecord);

/**
 * The base write schema for the "public.users" table.
 * This schema is used to validate the data before writing to the database without any transformations.
 */
export const UsersTableWriteSchema = z.object({
    /**
    * dataType: varchar
    * defaultValue: 
    */
    name: z.string().max(100),
    /**
    * dataType: varchar
    * defaultValue: 
    */
    email: z.string().max(255).nullish().optional(),
    /**
    * dataType: varchar
    * defaultValue: 'active'::character varying
    */
    status: z.string().max(20).nullish().optional(),
    /**
    * dataType: jsonb
    * defaultValue: 
    */
    profile: z.any().nullish().transform((value) => value ? JSON.stringify(value) : value).optional(),
    /**
    * dataType: timestamptz
    * defaultValue: now()
    */
    createdAt: z.date().nullish().optional(),
    /**
    * dataType: _text
    * defaultValue: 
    */
    roles: z.array(z.enum(USER_ROLES)).nullish().optional(),
    /**
    * dataType: _timestamp
    * defaultValue: 
    */
    dates: z.array(z.date()).nullish().optional(),
});

/**
 * The base record type for the "public.users" table.
 * This type represents an insertable database record before casing transformations are applied.
 */
export type UserInsertBaseRecord = z.output<typeof UsersTableWriteSchema>;

/**
 * The base record type for the "public.users" table.
 * This type represents an updatable database record before casing transformations are applied.
 */
export type UserUpdateBaseRecord = Partial<UserInsertBaseRecord>;

/**
 * The insert transform function for the "public.users" table.
 * Maps the expected property names to raw database fields. e.g camelCase to snake_case.
 */
export const transformUserInsertRecord = (data: UserInsertBaseRecord): {
    name: UserInsertBaseRecord['name'],
    email?: UserInsertBaseRecord['email'],
    status?: UserInsertBaseRecord['status'],
    profile?: UserInsertBaseRecord['profile'],
    created_at?: UserInsertBaseRecord['createdAt'],
    roles?: UserInsertBaseRecord['roles'],
    dates?: UserInsertBaseRecord['dates'],
} => ({
    name: data.name,
    email: data.email,
    status: data.status,
    profile: data.profile,
    created_at: data.createdAt,
    roles: data.roles,
    dates: data.dates,
});

/**
 * The update transform function for the "public.users" table.
 * Maps the expected property names to raw database fields. e.g camelCase to snake_case.
 */
export const transformUserUpdateRecord = (data: UserUpdateBaseRecord): {
    name: UserUpdateBaseRecord['name'],
    email?: UserUpdateBaseRecord['email'],
    status?: UserUpdateBaseRecord['status'],
    profile?: UserUpdateBaseRecord['profile'],
    created_at?: UserUpdateBaseRecord['createdAt'],
    roles?: UserUpdateBaseRecord['roles'],
    dates?: UserUpdateBaseRecord['dates'],
} => ({
    name: data.name,
    email: data.email,
    status: data.status,
    profile: data.profile,
    created_at: data.createdAt,
    roles: data.roles,
    dates: data.dates,
});

/**
 * The insert schema for the "public.users" table.
 * This schema is used to validate and transform a record before inserting into the database.
 */
export const UsersTableInsertSchema = UsersTableWriteSchema.transform(transformUserInsertRecord);

/**
 * The update schema for the "public.users" table.
 * This schema is used to validate and transform a record before updating the database.
 */
export const UsersTableUpdateSchema = UsersTableWriteSchema.partial().transform(transformUserUpdateRecord);

type TableInsertRecord = z.input<typeof UsersTableInsertSchema>;
type TableReadRecord = z.output<typeof UsersTableSchema>;
export type UserRole = (typeof USER_ROLES)[number];

/**
* Represents a database record from the "public.users" table.
*/
export interface UserRecord {
    /**
    */
    id: TableReadRecord['id'];
    /**
    */
    name: TableReadRecord['name'];
    /**
    */
    email?: TableReadRecord['email'];
    /**
    */
    status?: TableReadRecord['status'];
    /**
    */
    profile?: TableReadRecord['profile'];
    /**
    */
    createdAt?: TableReadRecord['createdAt'];
    /**
    */
    roles?: TableReadRecord['roles'];
    /**
    */
    dates?: TableReadRecord['dates'];
}

/**
* Represents an insertable database record from the "public.users" table.
*/
export interface UserInsertRecord {
    /**
    * @maxLen: 100
    */
    name: TableInsertRecord['name'];
    /**
    * @maxLen: 255
    */
    email?: TableInsertRecord['email'];
    /**
    * @maxLen: 20
    * @default: 'active'::character varying
    */
    status?: TableInsertRecord['status'];
    /**
    */
    profile?: TableInsertRecord['profile'];
    /**
    * @default: now()
    */
    createdAt?: TableInsertRecord['createdAt'];
    /**
    */
    roles?: TableInsertRecord['roles'];
    /**
    */
    dates?: TableInsertRecord['dates'];
}

/**
* Represents an updateable database record from the "public.users" table.
*/
export type UserUpdateRecord = Partial<UserInsertRecord>;
"
`;

exports[`module resolution formats > generates modules without file extensions (commonjs module resolution) > commonjs/types.ts 1`] = `
"// This file is auto-generated by zod-pg. Do not edit this file directly.

export type Table = 
    | 'users'
"
`;
